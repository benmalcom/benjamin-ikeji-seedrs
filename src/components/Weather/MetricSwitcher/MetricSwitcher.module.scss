.metric-switch-controls{
  margin-top: 5rem;
  width: 100%;
  @media (min-width: 320px) and (max-width: 480px) {
    margin-top: 2.5rem;
  }
}

.metric-switch-radio-group{
  display: flex;
  flex-direction: row !important;
  justify-content: center;
}

.switch-control{
  display: inline;
}