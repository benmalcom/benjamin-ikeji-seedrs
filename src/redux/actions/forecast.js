import { createActionType, createActionString } from './_create';

export const FETCH_FORECASTS = createActionType('FETCH_FORECASTS', 'Weather');
export const RESET_FORECAST_STATE = createActionType('RESET_FORECAST_STATE', 'Weather');
export const APPLY_METRICS_TO_FORECASTS = createActionString('APPLY_METRICS_TO_FORECASTS', 'Forecast');
export const UPDATE_FORECAST_LIST = createActionString('UPDATE_FORECAST_LIST', 'Forecast');
export const SET_CURRENT_METRIC = createActionString('SET_CURRENT_METRIC', 'Temperature');
export const REMOVE_FORECAST = createActionString('REMOVE_FORECAST', 'Forecast');

export const fetchForecasts = (params = {}) => ({
  type: FETCH_FORECASTS.START,
  meta: {
    params,
  },
});

export const resetForecastState = () => ({
  type: RESET_FORECAST_STATE.START,
});

export const applyMetricToForecasts = (metric) => ({
  type: APPLY_METRICS_TO_FORECASTS,
  meta: {
    metric,
  },
});

export const updateForecastList = (payload) => ({
  type: UPDATE_FORECAST_LIST,
  payload,
});

export const setCurrentMetric = (metric) => ({
  type: SET_CURRENT_METRIC,
  payload: { metric },
});

export const removeForecast = (id) => ({
  type: REMOVE_FORECAST,
  payload: { id },
});


