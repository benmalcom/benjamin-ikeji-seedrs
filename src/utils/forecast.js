export const metrics = {
  CELCIUS: 'Celcius',
  FAHRENHEIT: 'Fahrenheit',
};

export const metricValues = {
  CELCIUS: 'C',
  FAHRENHEIT: 'F',
};

export const metricValuesIconMap = {
  C: '\xB0C',
  F: '\xB0F',
};

export const celsiusToFahrenheit = value => value * 9 / 5 + 32;

export const fahrenheitToCelcius = value => (value - 32) * 5 / 9;

export const roundToTemp = value => Math.round(value);
