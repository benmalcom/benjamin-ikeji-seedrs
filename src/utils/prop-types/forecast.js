import PropTypes from 'prop-types';

export default {
  main: PropTypes.shape({
    temp: PropTypes.number.isRequired,
  }),
  weather: PropTypes.arrayOf(PropTypes.shape(
    {
      main: PropTypes.string.isRequired,
      icon: PropTypes.string.isRequired,
    }),
  ),
  dt_txt: PropTypes.string.isRequired,
  avg_temp: PropTypes.number,
};