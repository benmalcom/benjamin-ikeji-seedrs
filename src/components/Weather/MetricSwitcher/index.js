import React from 'react';
import { Box, FormControlLabel, RadioGroup } from '@material-ui/core';
import PropTypes from 'prop-types';
import styles from './MetricSwitcher.module.scss';
import StyledRadio from '../../UI/StyledRadio';
import { metrics, metricValues } from 'utils/forecast';

const MetricSwitcher = (props) => {
  const { onChange, value } = props;
  return <Box component='div' className={styles['metric-switch-controls']}>
    <RadioGroup className={styles['metric-switch-radio-group']} value={value} onChange={onChange} name='radio-button-metric'>
      <FormControlLabel
        role="radio"
        className={styles['switch-control']}
        value={metricValues.CELCIUS}
        control={<StyledRadio />}
        checked={value === metricValues.CELCIUS}
        label={metrics.CELCIUS} />
      <FormControlLabel
        role="radio"
        className={styles['switch-control']}
        value={metricValues.FAHRENHEIT}
        checked={value === metricValues.FAHRENHEIT}
        control={<StyledRadio />}
        label={metrics.FAHRENHEIT} />
    </RadioGroup>

  </Box>;
};

MetricSwitcher.propTypes = {
  onChange: PropTypes.func.isRequired,
  value: PropTypes.string,
};

MetricSwitcher.defaultProps = {
  value: null,
};

export default MetricSwitcher;