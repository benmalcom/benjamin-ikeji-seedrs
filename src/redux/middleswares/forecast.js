import { groupBy, last } from 'lodash';
import dayjs from 'dayjs';
import { v1 as uuid } from 'uuid';
import {
  apiRequest,
  APPLY_METRICS_TO_FORECASTS,
  FETCH_FORECASTS,
  updateForecastList,
  setCurrentMetric,
} from '../actions';
import { requestKeys } from 'services/_request/keys';
import { celsiusToFahrenheit, fahrenheitToCelcius, metricValues } from 'utils/forecast';

const fetchForecasts = ({ dispatch, getState }) => next => action => {
  next(action);
  if (action.type === FETCH_FORECASTS.START) {
    const { ui: { pagination } } = getState();
    const { key = requestKeys.fetchForecasts, params, ...rest } = action.meta;

    const defaultPagination = pagination.default;

    console.log("params ", params);

    dispatch(
      apiRequest({
        url: '/forecast',
        key: key,
        params: {
          ...params,
          ...defaultPagination,
        },
        onSuccess: data => {
          const payload = { city: data?.city?.name, id: uuid() };
          const byDate = groupBy(data?.list || [], item => dayjs(item.dt_txt).format('DD/MM/YYYY'));
          const forecasts = Object.values(byDate).map(groupList => last(groupList));
          payload.list = forecasts.length > 3 ? forecasts.slice(0, 3) : forecasts;
          if (!forecasts.length) {
            payload.message = 'No weather information about this city';
          }
          dispatch({ type: FETCH_FORECASTS.SUCCESS, payload });
        },
        onError: e => {
          const payload = {
            list: [],
            id: uuid(),
            message: Number(e?.cod) === 404 ? 'No weather information about this city' : 'Weather information not available',
            city: params?.q
          };
          dispatch({ type: FETCH_FORECASTS.ERROR, payload });
        },
        ...rest,
      }),
    );
  }
};

const applyMetricToForecasts = ({ dispatch, getState }) => next => action => {
  next(action);
  if (action.type === APPLY_METRICS_TO_FORECASTS) {
    const { metric } = action.meta;
    const { forecasts: { byList } } = getState();
    if (!metric || !byList?.length)
      return;
    const payload = byList.map(forecast => {
      const isToCelcius = metricValues.CELCIUS === metric;
      const metricFunc = isToCelcius ? fahrenheitToCelcius : celsiusToFahrenheit;

      return {
        ...forecast,
        list: forecast.list.map(item => ({
          ...item,
          main: {
            ...item.main,
            temp: metricFunc(item.main?.temp),
            feels_like: metricFunc(item.main?.feels_like),
            temp_min: metricFunc(item.main?.temp_min),
            temp_max: metricFunc(item.main?.temp_max),
          },
        })),
      };
    });
    dispatch(updateForecastList(payload));
    dispatch(setCurrentMetric(metric));
  }
};

export default [fetchForecasts, applyMetricToForecasts];
