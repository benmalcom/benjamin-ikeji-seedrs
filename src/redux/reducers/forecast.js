import {
  FETCH_FORECASTS,
  RESET_FORECAST_STATE,
  SET_CURRENT_METRIC,
  REMOVE_FORECAST,
  UPDATE_FORECAST_LIST
} from '../actions';
import { metricValues } from '../../utils/forecast';

const initialState = {
  currentMetric: metricValues.CELCIUS,
  byList: [],
};

const forecastReducer = (state = initialState, action) => {
  const { payload, type } = action;
  switch (type) {
    case FETCH_FORECASTS.ERROR:
    case FETCH_FORECASTS.SUCCESS: {
      return Object.assign({}, state, {
        ...state,
        byList: [payload,...state.byList]
      });
    }
    case SET_CURRENT_METRIC:
      return {
        ...state,
        currentMetric: payload.metric,
      };
    case UPDATE_FORECAST_LIST:
      return {
        ...state,
        byList: payload,
      };
    case REMOVE_FORECAST:
      return {
        ...state,
        byList: state.byList.filter(item => item.id !== payload.id),
      };
    case RESET_FORECAST_STATE:
      return initialState;
    default:
      return state;
  }
};

export default forecastReducer;

