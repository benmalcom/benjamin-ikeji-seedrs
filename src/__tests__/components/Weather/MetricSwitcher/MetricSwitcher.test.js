import React from 'react';
import { render } from '@testing-library/react';
import MetricSwitcher from 'components/Weather/MetricSwitcher';
import { metricValues } from 'utils/forecast';

describe('MetricSwitcher', () => {
  const handleChange = jest.fn();
  const value = metricValues.CELCIUS;
  const getComponent = props => render(<MetricSwitcher {...props} />);

  it('renders without crash', async () => {
    const component = getComponent({ value, onChange: handleChange });
    expect(component).toMatchSnapshot();
  });

});