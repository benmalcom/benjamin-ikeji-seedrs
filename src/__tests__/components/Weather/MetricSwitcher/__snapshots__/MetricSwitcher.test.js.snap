// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MetricSwitcher renders without crash 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiBox-root MuiBox-root-1 metric-switch-controls"
      >
        <div
          class="MuiFormGroup-root metric-switch-radio-group"
          role="radiogroup"
        >
          <label
            class="MuiFormControlLabel-root switch-control"
            role="radio"
          >
            <span
              aria-disabled="false"
              class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-5 MuiRadio-root makeStyles-root-2 PrivateSwitchBase-checked-6 Mui-checked"
            >
              <span
                class="MuiIconButton-label"
              >
                <input
                  checked=""
                  class="PrivateSwitchBase-input-8"
                  name="radio-button-metric"
                  type="radio"
                  value="C"
                />
                <span
                  class="makeStyles-icon-3 makeStyles-checkedIcon-4"
                />
              </span>
            </span>
            <span
              class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
            >
              Celcius
            </span>
          </label>
          <label
            class="MuiFormControlLabel-root switch-control"
            role="radio"
          >
            <span
              aria-disabled="false"
              class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-5 MuiRadio-root makeStyles-root-2"
            >
              <span
                class="MuiIconButton-label"
              >
                <input
                  class="PrivateSwitchBase-input-8"
                  name="radio-button-metric"
                  type="radio"
                  value="F"
                />
                <span
                  class="makeStyles-icon-3"
                />
              </span>
            </span>
            <span
              class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
            >
              Fahrenheit
            </span>
          </label>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiBox-root MuiBox-root-1 metric-switch-controls"
    >
      <div
        class="MuiFormGroup-root metric-switch-radio-group"
        role="radiogroup"
      >
        <label
          class="MuiFormControlLabel-root switch-control"
          role="radio"
        >
          <span
            aria-disabled="false"
            class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-5 MuiRadio-root makeStyles-root-2 PrivateSwitchBase-checked-6 Mui-checked"
          >
            <span
              class="MuiIconButton-label"
            >
              <input
                checked=""
                class="PrivateSwitchBase-input-8"
                name="radio-button-metric"
                type="radio"
                value="C"
              />
              <span
                class="makeStyles-icon-3 makeStyles-checkedIcon-4"
              />
            </span>
          </span>
          <span
            class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            Celcius
          </span>
        </label>
        <label
          class="MuiFormControlLabel-root switch-control"
          role="radio"
        >
          <span
            aria-disabled="false"
            class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-5 MuiRadio-root makeStyles-root-2"
          >
            <span
              class="MuiIconButton-label"
            >
              <input
                class="PrivateSwitchBase-input-8"
                name="radio-button-metric"
                type="radio"
                value="F"
              />
              <span
                class="makeStyles-icon-3"
              />
            </span>
          </span>
          <span
            class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            Fahrenheit
          </span>
        </label>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
